src/index.js

```diff
@@ -1,98 +1,140 @@
+ * @param {Array<number>} [options.minValues] - Minimum values for the parameters
+ * @param {number} [options.errorPropagation = 50] - How many evaluations (per point per step) of the fitted function to use to approximate the error propagation through it
+ * @return {Result}
  */
-export default function levenbergMarquardt(
-  data,
-  parameterizedFunction,
-  options = {}
-) {
+export default function levenbergMarquardt(data, paramFunction, options = {}) {
   let {
     maxIterations = 100,
-    gradientDifference = 10e-2,
-    damping = 0,
-    errorTolerance = 10e-3,
-    minValues,
+    gradientDifference = 1e-6,
+    damping = 0.1,
+    dampingDrop = 0.1,
+    dampingBoost = 1.5,
+    maxDamping = Number.MAX_SAFE_INTEGER,
+    minDamping = Number.EPSILON,
     maxValues,
-    initialValues
-  } = options;
+    minValues,
+    residualEpsilon = 1e-6,
+    initialValues,
+    errorPropagation = 50
+  } = legacy.compatOptions(options);
 
   if (damping <= 0) {
     throw new Error('The damping option must be a positive number');
-  } else if (!data.x || !data.y) {
-    throw new Error('The data parameter must have x and y elements');
+  } else if (!data || !data.x || !data.y) {
+    throw new Error('The data object must have x and y elements');
   } else if (
     !Array.isArray(data.x) ||
     data.x.length < 2 ||
     !Array.isArray(data.y) ||
     data.y.length < 2
   ) {
+    throw new Error('The data must have more than 2 points');
+  } else if (data.x.length !== data.y.length) {
     throw new Error(
-      'The data parameter elements must be an array with more than 2 points'
+      'The data object must have equal number of x and y coordinates'
     );
-  } else if (data.x.length !== data.y.length) {
-    throw new Error('The data parameter elements must have the same size');
   }
 
-  var parameters =
-    initialValues || new Array(parameterizedFunction.length).fill(1);
-  let parLen = parameters.length;
+  let params = initialValues || new Array(paramFunction.length).fill(1);
+  let parLen = params.length;
   maxValues = maxValues || new Array(parLen).fill(Number.MAX_SAFE_INTEGER);
   minValues = minValues || new Array(parLen).fill(Number.MIN_SAFE_INTEGER);
 
-  if (maxValues.length !== minValues.length) {
-    throw new Error('minValues and maxValues should have the same size');
+  if (!Array.isArray(params)) {
+    throw new Error('initialValues must be an array');
   }
 
-  if (!Array.isArray(parameters)) {
-    throw new Error('initialValues must be an array');
+  if (
+    maxValues.length !== minValues.length ||
+    maxValues.length !== params.length
+  ) {
+    throw new Error(
+      'minValues and maxValues must have the same size as the number of parameters'
+    );
   }
 
-  var error = errorCalculation(data, parameters, parameterizedFunction);
+  initializeErrorPropagation(errorPropagation);
+
+  /** @type Array<number> */
+  var residualDifferences = Array(10).fill(NaN);
```

Comment:
What is the significance of `10` here? It seems like a ["magic number"](http://stackoverflow.com/questions/47882/ddg#47902)

Reply:
Am I correct in interpreting the intention here as: (1) require at least 10 iterations and (2) stop after the total amount of change in the last 10 residuals is less than `residualEpsilon`?

Reply:
Well, I'm not saying that it should be configurable (maybe it should, maybe not; I don't know yet). I'm saying that the code should be more [expressive](https://medium.com/@daniel.oliver.king/writing-expressive-code-b69ef7a5a2fa) (Note: this is just a free article I found from searching the web; there are much better resources on this topic, such as the book "Clean Code" by Robert C. Martin). In other words, when a developer looks at this code, he or she should not have to ask "What does this number mean?" -- the answer should be already provided by giving it a name.

Reply:
In fact, I would even suggest taking it a step further to cover the question of "Why is this array being filled with `NaN`"? However, I don't think the solution to that problem lies in improving the name but rather in refactoring the code. It seems like we are only concerned about (1) whether or not we have taken at least 10 steps and (2) whether or not the biggest change in the last 10 steps was smaller than some given threshold. So, for example, we could do something like this instead of having an array, filling it with `NaN`, shifting it, scanning it, etc.

```js
// Outside of loop
const minIterations = 10;   // We won't use the `residualEpsilon` stopping condition until it's been stable for this many steps
let maxEpsilon = -Infinity; // This will get replaced during the first step because the new values must be > -Infinity (alternatively we could set to null or undefined and handle that as a special case
let iterationsRemaining;    // This count-down will get initialized during the first step because we initialize maxEpsilon to its minimum
// Inside of loop
  // ... algorithm takes a step ...
  let currEpsilon = prevResiduals - currResiduals; // (may need ensure that this value is >= 0)
  if (currEpsilon > maxEpsilon) {
    // We just found a new maximum, so we reset the count-down before allowing this to be used as a stopping condition
    // This eliminates the need to maintain an array of previous epsilon values
    maxEpsilon = maxEpsilon;
    iterationsRemaining = minIterations;
  }
  else {
    // maxEpsilon did not change so we can decrement our counter (limiting to zero) 
    iterationsRemaining--;
    if (iterationsRemaining < 0) {
      iterationsRemaining = 0;
    }	
  }
  const isChangeTooSmall = maxEpsilon <= residualEpsilon; 
  if (isChangeTooSmall && iterationsRemaining === 0) {
    // ... stop ...
  }
```